File: services/project-management-service/src/main/java/com/example/project_management_service/model/Project.java
package com.example.projectmanagementservice.model;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "projects")
public class Project {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private UUID id;

    @Column(nullable = false, length = 55)
    private String title;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String description;

    @Column(nullable = false, length = 50)
    private String category;

    @Column(nullable = false)
    private BigDecimal price;

    @Column(nullable = false)
    private boolean allowHigherPrice;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Column(nullable = false)
    private Long userId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProjectStatus status;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Getters and setters

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public boolean isAllowHigherPrice() {
        return allowHigherPrice;
    }

    public void setAllowHigherPrice(boolean allowHigherPrice) {
        this.allowHigherPrice = allowHigherPrice;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Long getUserId() {
        return userId;
    }

    public void setUserId(Long userId) {
        this.userId = userId;
    }

    public ProjectStatus getStatus() {
        return status;
    }

    public void setStatus(ProjectStatus status) {
        this.status = status;
    }
}

File: services/search-service/internal/model/project.go
package model

import (
    "time"
    "github.com/google/uuid"
    "gorm.io/gorm"
)

type ProjectStatus string

const (
    StatusCreated    ProjectStatus = "CREATED"
    StatusInProgress ProjectStatus = "IN_PROGRESS"
    StatusPaused     ProjectStatus = "PAUSED"
    StatusCompleted  ProjectStatus = "COMPLETED"
)

type Project struct {
    ID              uuid.UUID      `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
    Title           string         `gorm:"type:varchar(55);not null" json:"title"`
    Description     string         `gorm:"type:text;not null" json:"description"`
    Category        string         `gorm:"type:varchar(50);not null" json:"category"`
    Price           float64        `gorm:"type:numeric;not null" json:"price"`
    AllowHigherPrice bool          `gorm:"not null" json:"allow_higher_price"`
    CreatedAt       time.Time      `gorm:"not null;autoCreateTime" json:"created_at"`
    UpdatedAt       time.Time      `gorm:"not null;autoUpdateTime" json:"updated_at"`
    UserID          uint           `gorm:"not null" json:"user_id"`
    Status          ProjectStatus  `gorm:"type:varchar(20);not null" json:"status"`
}

func (p *Project) BeforeCreate(tx *gorm.DB) (err error) {
    p.ID = uuid.New()
    return
}

func (p *Project) BeforeUpdate(tx *gorm.DB) (err error) {
    p.UpdatedAt = time.Now()
    return
}
